frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,139 @@ use frame_support::traits::{
 use frame_system::{ensure_signed, ensure_root, pallet_prelude::*};
 pub use weights::WeightInfo;
 pub use pallet::*;
+pub use vesting_info::*;
 
 type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
 type MaxLocksOf<T> = <<T as Config>::Currency as LockableCurrency<<T as frame_system::Config>::AccountId>>::MaxLocks;
 
 const VESTING_ID: LockIdentifier = *b"vesting ";
 
-/// Struct to encode the vesting schedule of an individual account.
-#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
-pub struct VestingInfo<Balance, BlockNumber> {
-	/// Locked amount at genesis.
-	pub locked: Balance,
-	/// Amount that gets unlocked every block after `starting_block`.
-	pub per_block: Balance,
-	/// Starting block for unlocking(vesting).
-	pub starting_block: BlockNumber,
+// Module to enforce private fields on `VestingInfo`
+mod vesting_info {
+	use super::*;
+	/// Struct to encode the vesting schedule of an individual account.
+	#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
+	pub struct VestingInfo<Balance, BlockNumber> {
+		/// Locked amount at genesis.
+		locked: Balance,
+		/// Amount that gets unlocked every block after `starting_block`.
+		per_block: Balance,
+		/// Starting block for unlocking(vesting).
+		starting_block: BlockNumber,
+	}
+
+	impl<Balance: AtLeast32BitUnsigned + Copy, BlockNumber: AtLeast32BitUnsigned + Copy>
+		VestingInfo<Balance, BlockNumber>
+	{
+		/// Instantiate a new `VestingInfo` and validate parameters
+		pub fn try_new<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> Result<VestingInfo<Balance, BlockNumber>, Error<T>> {
+			Self::validate_params(locked, per_block, starting_block)?;
+			let per_block = if per_block > locked { locked } else { per_block };
+			Ok(VestingInfo { locked, per_block, starting_block })
+		}
+
+		/// Validate parameters for `VestingInfo`.
+		pub fn validate_params<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			_starting_block: BlockNumber,
+		) -> Result<(), Error<T>> {
+			ensure!(!locked.is_zero() && !per_block.is_zero(), Error::<T>::InvalidScheduleParams);
+			let min_transfer: u32 = T::MinVestedTransfer::get().try_into().unwrap_or(u32::MAX);
```

Comment:
It looks like the intent here is that if `T::MinVestedTransfer::get()` returns `None`, then we make the minimum into `u32::MAX` so that the subsequent `ensure` always returns `AmountLow`. 

However, it's not uncommon for currencies to be denominated in much wider types, and it's entirely plausible that some user will entirely innocently lock a value greater than `u32::MAX` entirely by accident.

If I've read the intent correctly, then it would be better to write this as

```rust
let min_transfer = T::MinVestedTransfer::try_get().ok_or(Error::<T>::TransferDisabled)?;
```

(with appropriate bikeshedding for the error type) instead of going through the `u32` conversion.

If I've _not_ read the intent correctly, then this would be a good place for a detailed comment explaining what you meant, and why you picked `u32::MAX`.

Reply:
yeah i think i need to get rid of this conversion, they should both be `Balance` so i should not need to convert

Reply:
I think this is being checked too high in the API.

The `MinVestedTransfer` should only apply for specifically the extrinsic `vested_transfer`

Any other creation of a schedule does not have these requirements